transports/websocket/src/framed.rs

```diff
@@ -98,6 +98,31 @@ impl<T> WsConfig<T> {
         self.use_deflate = flag;
         self
     }
+
+    /// Look backward to if addr has a trailing "/ws" and returns a Wss protocol if "/tls" was found.
+    pub(crate) fn get_address_proto(
```

Comment:
Does this have to be on `Self`? Could be a free function :)

Reply:
Also, we could refactor this to not mutate its argument and instead consume the argument and return a new address.

Reply:
And by not mutating its argument and returning a new address, do you mean the function also returns another `Multiaddr` back to the caller? i.e. something like

```rust
fn get_address_proto(addr: Multiaddr) -> Result<(bool, Protocol<'static>, Multiaddr), String>
```

Reply:
> Consuming the Multiaddr will likely cause one additional clone

Looking at the current usage, it is already cloned before being passed in or am I missing something?

The reasoning was: If the caller is okay with mutation (which it currently seems to be), then they can just pass over ownership and use the new return value so it shouldn't make a difference but a non-mutable API would be easier to understand IMO.

Reply:
> You mean a public function? Nothing private about it, just feels like it's a helper rather than an API.

Uhm, still pub(crate) or whatever it needs to be but just not on `Self` and instead free-standing (outside an `impl` block).

Reply:
How's this? [1e2117a](https://github.com/libp2p/rust-libp2p/pull/2487/commits/1e2117ad5a7046b6123f34e627851d075169d708)

Reply:
Yeah that is much easier to understand IMO, thank you :)